src/Controller/Admin/AddonsController.php

```diff
@@ -0,0 +1,150 @@
+        RequestStack $requestStack,
+        DataProvider $addonsDataProvider,
+        ModuleInterface $modulesDataProvider
+    ) {
+        parent::__construct();
+        $this->requestStack = $requestStack;
+        $this->addonsDataProvider = $addonsDataProvider;
+        $this->modulesDataProvider = $modulesDataProvider;
+    }
+
+    /**
+     * Controller responsible for the authentication on PrestaShop Addons.
+     *
+     * @return JsonResponse
+     */
+    public function loginAction(): JsonResponse
+    {
+        $request = $this->requestStack->getCurrentRequest();
+        $response = new JsonResponse();
+
+        // Parameters needed in order to authenticate the merchant : login and password
+        $params = [
+            'format' => 'json',
+            'username' => $request->get('username_addons', null),
+            'password' => $request->get('password_addons', null),
+        ];
+
+        try {
+            $json = $this->addonsDataProvider->request('check_customer', $params);
+            if ($json === null) {
+                throw new LoginErrorException();
+            }
+
+            if (!empty($json->errors)) {
+                throw new LoginErrorException($json->errors->code . ': ' . $json->errors->label);
+            }
+
+            Configuration::updateValue('PS_LOGGED_ON_ADDONS', 1);
+
+            $phpEncryption = new PhpEncryption(_NEW_COOKIE_KEY_);
+
+            $response->headers->setCookie(
+                new Cookie('username_addons', $phpEncryption->encrypt($params['username']))
+            );
+            $response->headers->setCookie(
+                new Cookie('password_addons', $phpEncryption->encrypt($params['password']))
+            );
+            $response->headers->setCookie(
+                new Cookie('is_contributor', (string) $json->is_contributor)
+            );
+
+            $response->setData(['success' => 1, 'message' => '']);
+            $this->modulesDataProvider->clearCatalogCache();
+        } catch (Exception $e) {
+            $response->setData([
+                'success' => 0,
+                'message' => $this->trans(
+                    'PrestaShop was unable to log in to Addons. Please check your credentials and your Internet connection.',
+                    'Admin.Notifications.Error'
+                ),
+            ]);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Controller responsible for the logout on PrestaShop Addons.
+     *
+     * @return JsonResponse|RedirectResponse
+     */
+    public function logoutAction()
+    {
+        $this->modulesDataProvider->clearCatalogCache();
+        $request = $this->requestStack->getCurrentRequest();
+
+        if ($request->isXmlHttpRequest()) {
+            $response = new JsonResponse();
+            $response->setData([
+                'success' => 1,
+                'message' => '',
+            ]);
+        } else {
+            if ($request->server->get('HTTP_REFERER')) {
```

Comment:
Watch out with HTTP_REFERER, you can forge a request easily: https://web.dev/referrer-best-practices/

Reply:
The alternatives given are only acting on server side. Do you have a recommandation/advice ?